libraries/reliable_amqp_publisher/reliable_amqp_publisher.cpp

```diff
@@ -0,0 +1,247 @@
+      AMQP::LibBoostAsioHandler(ctx),
+      cbs(callbacks) {}
+
+   void onReady(AMQP::TcpConnection* c) override {
+      cbs->amqp_ready();
+   }
+
+   //disable heartbeat for now
+   uint16_t onNegotiate(AMQP::TcpConnection* connection, uint16_t interval) override {
+      return 0;
+   }
+
+   void onError(AMQP::TcpConnection*, const char* message) override {
+      cbs->amqp_error(message);
+   }
+
+   reliable_amqp_publisher_callbacks* cbs;
+};
+
+struct reliable_amqp_publisher_impl final : reliable_amqp_publisher_callbacks {
+   reliable_amqp_publisher_impl(const std::string url, const std::string exchange, const std::string routing_key, const boost::filesystem::path unconfirmed_path, const std::optional<std::string> message_id) :
+     data_file_path(unconfirmed_path), exchange(exchange), routing_key(routing_key), message_id(message_id) {
+      std::string host = "localhost", user = "guest", pass = "guest", path;
+      uint16_t port = 5672;
+      fc::url parsed_url(url);
+      FC_ASSERT(parsed_url.proto() == "amqp", "Only amqp:// URIs are supported for AMQP addresses");
+      if(parsed_url.host() && parsed_url.host()->size())
+         host = *parsed_url.host();
+      if(parsed_url.user())
+         user = *parsed_url.user();
+      if(parsed_url.pass())
+         pass = *parsed_url.pass();
+      if(parsed_url.port())
+         port = *parsed_url.port();
+      if(parsed_url.path())
+         path = parsed_url.path()->string();
+
+      std::stringstream hostss;
+      hostss << "amqp://" << user << ":" << pass << "@" << host << ":" << port << "/" << path;
+      amqp_address = std::make_unique<AMQP::Address>(hostss.str());
+
+      if(boost::filesystem::exists(data_file_path)) {
+         try {
+            fc::datastream<fc::cfile> file;
+            file.set_file_path(data_file_path);
+            file.open("rb");
+            fc::raw::unpack(file, message_deque);
+         } FC_RETHROW_EXCEPTIONS(error, "Failed to load previously unconfirmed AMQP messages from ${f}", ("f", (fc::path)data_file_path));
+      }
+      else {
+         boost::filesystem::ofstream o(data_file_path);
+         FC_ASSERT(o.good(), "Failed to create unconfirmed AMQP message file at ${f}", ("f", (fc::path)data_file_path));
+      }
+      boost::system::error_code ec;
+      boost::filesystem::remove(data_file_path, ec);
+
+      thread = std::thread([this]() {
+         try {
+            fc::set_os_thread_name("amqp");
+            retry_connection();
+            ctx.run();
+         }
+         catch(...) {
+            elog("Exception escaped from AMQP connection ${s} publishing to \"${e}\", this connection is permanently disabled now",
+                 ("s", (std::string)*amqp_address)("e", this->exchange));
+         }
+      });
+   }
+
+   ~reliable_amqp_publisher_impl() {
+      stopping = true;
+
+      //drain any remaining items on queue
+      std::promise<void> shutdown_promise;
+      ctx.post([&]() {
```

Comment:
Maybe not the best comment. It's meant to drain everything submitted before this post(). In particular this means anything someone has submitted to `post_on_io_context()` has been run. That closes a small hole in time where someone could have submitted work to `post_on_io_context()` but an amqp message had not been created for it (and tucked away safely). Consider the witness plugin which submits something on to this queue that could take 15-100ms for a hardware key to complete.

Reply:
I can't find any guarantee that `post`s will be executed in order.

Reply:
hm, yeah, neither can I. Can rework to strand pretty easily.

Reply:
Actually there is some evidence here, https://www.boost.org/doc/libs/1_73_0/doc/html/boost_asio/overview/core/strands.html

> Calling io_context::run() from only one thread means all event handlers execute in an implicit strand

and then of course https://www.boost.org/doc/libs/1_73_0/doc/html/boost_asio/reference/io_context__strand.html is very clear that

>  `s.post(a)` happens-before `s.post(b)`

Reply:
Unfortunately, `s.post(a)` is an explicit strand, not an implicit one.  From reading the former, my impression is that an implicit strand does not provide any guarantees.  It's only a convenient way to describe how other synchronization rules can be combined to meet the definition of a strand.

Reply:
This still doesn't work if the user's event handler submits additional work to the queue.  In particular, `publish_message` calls `ctx.dispatch` for which "the handler __may__ be executed inside this function if the guarantee can be met." (emphasis added)

Reply:
io_contexts's executor makes stronger guarantees.
> If the current thread is running the io_context, dispatch() executes the function before returning. Otherwise, the function will be scheduled to run on the io_context.

Reply:
Though maybe to be perfectly pedantically in line with the documentation (and not use something depreciated anyways), the free `dispatch()` function should be used instead? that is explicitly documenting calling io_context's executor's `dispatch()`. The `io_context::dispatch()` is more ambiguous.

Reply:
Agreed.  On top of that, maybe add a note that the handler passed to `post_on_io_context` should not call `post_on_io_context`.

Reply:
I think there is another problem here where if `publish_message()` is called directly (not through `post_on_io_context()`) that dispatch() becomes a post() on the io_context and thus loses ordering and proper shut down semantics. I think this can be resolved fairly easily..